server-spi-private/src/main/java/org/keycloak/crypto/SignatureProvider.java

```diff
@@ -17,6 +17,7 @@
 package org.keycloak.crypto;
 
 import org.keycloak.common.VerificationException;
+import org.keycloak.jose.jwk.JWK;
 import org.keycloak.provider.Provider;
 
 public interface SignatureProvider extends Provider {
```

Comment:
How about using `ClientSignatureVerifierProvider` that is used for keycloak to verify the data that a client creates and signs (e.g. request object)?

`SignatureProvider` is used for keycloak to sign and verify JWS (e.g. access token, ID token, refresh token). DPoP Proof is the data that a client creates and signs and keycloak verify.

Reply:
When looking for an SPI in Keycloak that would allow to validate RSA/EC JWSs in a unified manner, I've ended up with these:
- `org.keycloak.crypto.ClientSignatureVerifierProvider`
- `org.keycloak.jose.jws.crypto.*Provider`
- `org.keycloak.crypto.SignatureProvider`

`ClientSignatureVerifierProvider` and friends would work well, but the key point about this SPI is that it recognizes only the keys that has been registered to the clients beforehand (hence the "Client" bit in the class names). OTOH, DPoP (whilst being initiated from the client side) uses ephemeral keys generated on the fly that Keycloak has no knowledge of. In theory, we could add methods taking just JWS and public key, but IMO this will result even in more changes to the whole hierarchy.

`org.keycloak.jose.jws.crypto.*Provider` classes are used in `TokenVerifier::verifySignature` in case no custom verifier has been supplied. I'm not sure if this code path is used in Keycloak at all, since e.g. there is no EC provider implementation in `org.keycloak.jose.jws.crypto`. In theory, we could add one, which will be yet third signature verification SPI alongside the other two.

With the above, I've considered `SignatureProvider` the most appropriate SPI. Yet, it required addition of `SignatureProvider::verifier(JWK key)` which had repercussions across the hierarchy.

I'm open to suggestions; which of the above SPIs seem most relevant to you?

Reply:
@dteleguin Thank you for the clarification. It might be better to choose a last option as you've implemented originally.